### PR TITLE
CLOUDSTACK-9453: WIP : Marvin optimizations and fixes.

### PR DESCRIPTION
**Pull request by abhinandanprateek**
1. Adding small sized macchinina template as built in templates
  2. Adding methods so the test writers haev choice to use any of these
  3. Updateing test_host_ha to use macchinina template
  4. Misc fixes encoutered while running component tests

### CODE REVIEW

**Comment by jburwell** [comment]
@abhinandanprateek the Jenkins build failed.  Please investigate and fix the error.

File: setup/db/db/schema-490to491.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
+LOCK TABLES `vm_template` WRITE;
+INSERT INTO `vm_template` VALUES (13,'macchinina-x86_64-xen','Macchinina (64-bit) no GUI (XenServer)','9511a500-620b-11e6-843c-00505635944c',1,1,'BUILTIN',0,64,'http://dl.openvm.eu/cloudstack/macchinina/x86_64/macchinina-xen.vhd.bz2','VHD','2016-08-14 16:11:00',NULL,1,'30985504bc31bf0cd3b9d2c6ca7944d3','Macchinina(64-bit) no GUI (XenServer)',0,0,99,1,0,1,1,'XenServer',NULL,NULL,0,NULL,'Active',0,NULL,0),(14,'macchinina-x86_64','Macchinina (64-bit) no GUI (KVM)','9512a500-620b-11e6-843c-00505635944c',1,1,'BUILTIN',0,64,'http://dl.openvm.eu/cloudstack/macchinina/x86_64/macchinina-kvm.qcow2.bz2','QCOW2','2016-08-14 16:11:02',NULL,1,'f50acb3a4387019b473d0f25a42bf06e','Macchinina (64-bit) no GUI (KVM)',0,0,99,1,0,1,1,'KVM',NULL,NULL,0,NULL,'Active',0,NULL,0),(15,'macchinina-x64','Macchinina (64-bit) no GUI (vSphere)','9513a500-620b-11e6-843c-00505635944c',1,1,'BUILTIN',0,64,'http://dl.openvm.eu/cloudstack/macchinina/x86_64/macchinina-vmware.ova','OVA','2016-08-14 16:11:05',NULL,1,'f6f881b7f2292948d8494db837fe0f47','Macchinina (64-bit) no GUI (vSphere)',0,0,99,1,0,1,1,'VMware',NULL,NULL,0,NULL,'Active',0,NULL,0);
+UNLOCK TABLES;
```
Comment (jburwell) [inline setup/db/db/schema-490to491.sql]: Why is a database change required for a Marvin optimization?

File: engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java [hunk 2]
```diff
@@ -92,6 +92,7 @@ public ObjectInDataStoreManagerImpl() {
         stateMachines.addTransition(State.Allocated, Event.OperationFailed, State.Failed);
         stateMachines.addTransition(State.Creating, Event.OperationFailed, State.Allocated);
         stateMachines.addTransition(State.Creating, Event.OperationSuccessed, State.Ready);
+        stateMachines.addTransition(State.Creating, Event.DestroyRequested, State.Destroying);
```
Comment (jburwell) [inline engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java]: Why is this change required for the Marvin optimization?

File: tools/marvin/marvin/lib/common.py [hunk 3]
```diff
@@ -334,13 +335,57 @@ def get_template(
...
     for template in list_templatesout:
-        if template.isready and template.templatetype == template_type:
+        if template.isready and template.templatetype == template_type and template.name.startswith('CentOS'):
             return template
     '''
     Return default first template, if no template matched
     '''
     return list_templatesout[0]
 
+def get_template_macchinina(
```
Comment (jburwell) [inline tools/marvin/marvin/lib/common.py]: This function and the ``get_template`` function seem very similar.  The only difference appears to be that  ``get_template`` function is checking for ``CentOS`` and ``get_template_macchinia`` is checking for ``macchinia``.  Please refactor these functions to remove the duplication.

File: tools/marvin/marvin/sshClient.py [hunk 4]
```diff
@@ -146,9 +146,8 @@ def createConnection(self):
                 if self.retryCnt == 0 or ret == SUCCESS:
                     break
                 if except_msg != '':
-                    self.logger.\
-                        exception("SshClient: Exception under "
-                                  "createConnection: %s" % except_msg)
+                    self.logger.debug("SshClient: Exception under "
+                                  "createConnection: %s while connecting to " % except_msg, self.host)
                 self.retryCnt -= 1
                 time.sleep(self.delay)
```
Comment (jburwell) [inline tools/marvin/marvin/sshClient.py]: Recognizing that this comment is beyond the scope of this PR, would it be possible to refactor lines 146-152 to use ``wait_for`` and a lambda?

File: setup/db/db/schema-490to491.sql (hunk 1 above)
Comment (abhinandanprateek) [inline setup/db/db/schema-490to491.sql]: This change will add xen macchinina template as builtin templates to cloudstack. This will ensure that these templates are available by default like the Centos templates.

File: engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java (hunk 2 above)
Comment (abhinandanprateek) [inline engine/storage/src/org/apache/cloudstack/storage/datastore/ObjectInDataStoreManagerImpl.java]: This is actually a bug that was hit while running Marvin. A Vm that is being created threw exception when destroyed as there is not respective state transition.

**Comment by jburwell** [comment]
@abhinandanprateek could you please investigate the cause of the Jenkins failure?

**Comment by jburwell** [comment]
@abhinandanprateek ping regarding the Travis and Jenkins failures.  I would like to include this PR in the upcoming 4.8.2.0, 4.9.1.0, and 4.10.0.0 releases for which we plan to cut an RC on 25 Sept 2016.

**Comment by abhinandanprateek** [comment]
@jburwell rebased it 

@rhtyd is there a way to move a PR to another branch without creating a new one ?

**Comment by karuturi** [comment]
@abhinandanprateek https://github.com/blog/2224-change-the-base-branch-of-a-pull-request

**Comment by rhtyd** [comment]
@abhinandanprateek yes, click on `edit` on your PR you should be able to change the base branch.

**Comment by abhinandanprateek** [comment]
Thanks @karuturi @rhtyd 
I created a new one for 4.8 here https://github.com/apache/cloudstack/pull/1675 @jburwell

**Comment by serg38** [comment]
Macchinina template doesn't seem to work in vmware hypervisor. When VM deployed eth0 is not created. We probably need to make sure a template is fully operational on all 3 hypervisors to include it as a default.

**Comment by abhinandanprateek** [comment]
@serg38 will test and see it this can be fixed.
